src/services/codefixes/convertTypedefToType.ts

```diff
@@ -0,0 +1,232 @@
+    TypeAliasDeclaration, ParenthesizedTypeNode, UnionTypeNode, PropertySignature,
+    TypeNodeSyntaxKind, TypeNode, HasJSDoc, isJSDocTypedefTag, hasJSDocNodes,
+    isInJSDoc, getTokenAtPosition, textChanges, flatMap
+} from "../_namespaces/ts";
+import { codeFixAll, createCodeFixAction, registerCodeFix } from "../_namespaces/ts.codefix";
+
+const fixId = "convertTypedefToType";
+const errorCodes = [Diagnostics.Convert_typedef_to_type.code];
+registerCodeFix({
+    fixIds: [fixId],
+    errorCodes,
+    getCodeActions(context) {
+        const node = getTokenAtPosition(
+            context.sourceFile,
+            context.span.start
+        );
+        if (!node) return;
+        const changes = textChanges.ChangeTracker.with(context, t => doChange(t, node, context.sourceFile));
+
+        if (changes.length > 0) {
+            return [
+                createCodeFixAction(
+                    fixId,
+                    changes,
+                    Diagnostics.Convert_typedef_to_type,
+                    fixId,
+                    Diagnostics.Convert_all_typedefs_to_types
+                ),
+            ];
+        }
+    },
+    getAllCodeActions: context => codeFixAll(context, errorCodes, (changes, diag) => {
+        const node = getTokenAtPosition(diag.file, diag.start);
+        if (node) doChange(changes, node, diag.file);
+    })
+});
+
+function doChange(changes: textChanges.ChangeTracker, node: Node, sourceFile: SourceFile) {
+    if (containsTypeDefTag(node)) {
+        fixSingleTypeDef(changes, node, sourceFile);
+    }
+}
+
+function fixSingleTypeDef(
+    changes: textChanges.ChangeTracker,
+    typeDefNode: JSDocTypedefTag | undefined,
+    sourceFile: SourceFile,
+) {
+    if (!typeDefNode) return;
+
+    const declaration = createDeclaration(typeDefNode);
+    if(!declaration) return;
+
+    const comment = typeDefNode.parent;
+
+    changes.replaceNode(
+        sourceFile,
+        comment,
+        declaration
+    );
+}
+
+function createDeclaration(tag: JSDocTypedefTag): InterfaceDeclaration | TypeAliasDeclaration | undefined {
+    const { typeExpression } = tag;
+    if (!typeExpression) return;
+    const typeName = tag.name?.getFullText().trim();
+    if (!typeName) return;
+
+    // For use case @typedef {object}Foo @property{bar}number
+    // But object type can be nested, meaning the value in the k/v pair can be object itself
+    if (typeExpression.kind === SyntaxKind.JSDocTypeLiteral) {
+        return createInterfaceForTypeLiteral(typeName, typeExpression);
+    }
+    // for use case @typedef {(number|string|undefined)} Foo or @typedef {number|string|undefined} Foo
+    // In this case, we reach the leaf node of AST.
+    // Here typeExpression.type is a TypeNode, e.g. a UnionType or a primitive type, such as "number".
+    if (typeExpression.kind === SyntaxKind.JSDocTypeExpression) {
+     return createTypeAliasForTypeExpression(typeName, typeExpression);

```

Comment:
nit: bad indent

There are actually a lot of these spacing problems in the file; I'll stop commenting on them as I wish we had a formatter to prevent this, but I would take a look at the diff and just see what's missing.

Reply:
fixed